test/framework/src/main/java/org/elasticsearch/test/gateway/TestGatewayAllocator.java

```diff
@@ -0,0 +1,142 @@
+                .filter(ks -> ignoreNodes.contains(ks.currentNodeId()) == false)
+                .filter(ks -> knownNodes.containsKey(ks.currentNodeId()))
+                .collect(Collectors.toMap(
+                    routing -> knownNodes.get(routing.currentNodeId()),
+                    routing ->
+                        new NodeGatewayStartedShards(
+                            knownNodes.get(routing.currentNodeId()), -1, routing.allocationId().getId(), routing.primary())));
+
+            return new AsyncShardFetch.FetchResult<>(shardId, foundShards, Collections.emptySet(), ignoreNodes);
+        }
+    };
+
+    ReplicaShardAllocator replicaShardAllocator = new ReplicaShardAllocator(Settings.EMPTY) {
+        @Override
+        protected AsyncShardFetch.FetchResult<NodeStoreFilesMetaData> fetchData(ShardRouting shard, RoutingAllocation allocation) {
+            // for now, just pretend no node has data
+            final ShardId shardId = shard.shardId();
+            return new AsyncShardFetch.FetchResult<>(shardId, Collections.emptyMap(), Collections.emptySet(),
+                allocation.getIgnoreNodes(shardId));
+        }
+    };
+
+    public TestGatewayAllocator() {
+        super(Settings.EMPTY, null, null);
+    }
+
+    @Override
+    public void applyStartedShards(StartedRerouteAllocation allocation) {
+        updateNodes(allocation);
+        for (ShardRouting shard: allocation.routingNodes().shards(ShardRouting::active)) {
+            addKnownAllocation(shard);
+        }
+    }
+
+    @Override
+    public void applyFailedShards(FailedRerouteAllocation allocation) {
+        updateNodes(allocation);
+        for (FailedRerouteAllocation.FailedShard failedShard : allocation.failedShards()) {
+            final ShardRouting failedRouting = failedShard.routingEntry;
+            Map<ShardId, ShardRouting> nodeAllocations = knownAllocations.get(failedRouting.currentNodeId());
+            if (nodeAllocations != null) {
+                nodeAllocations.remove(failedRouting.shardId());
+                if (nodeAllocations.isEmpty()) {
+                    knownAllocations.remove(failedRouting.currentNodeId());
+                }
+            }
+        }
+    }
+
+    private void updateNodes(RoutingAllocation allocation) {
+        knownNodes.clear();
+        for (DiscoveryNode node: allocation.nodes()) {
+            knownNodes.put(node.getId(), node);
+        }
+    }
+
+    @Override
+    public void allocateUnassigned(RoutingAllocation allocation) {
+        updateNodes(allocation);
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
+        unassigned.sort(PriorityComparator.getAllocationComparator(allocation)); // sort for priority ordering
+
+        primaryShardAllocator.allocateUnassigned(allocation);
+        replicaShardAllocator.processExistingRecoveries(allocation);
+        replicaShardAllocator.allocateUnassigned(allocation);
```

Comment:
I now remember why I did it this way - I couldn't find a good way to get all the inner classes to work as they can not be constructed and passed around as a parameter. I took another approach with https://github.com/elastic/elasticsearch/pull/20637/commits/13b5c5c3272aedbc631d578d1e585886bef817da